penney/functions.py

```diff
@@ -98,6 +98,89 @@ def find_winner(seq, seq_dict):
     return None
 
 
+def det(A):
+    """
+    Calculate determinant of a matrix in a fast way.
+
+    :param A: matrix itself
+    :type A: list or numpy.array
+    :return: determinant of A in float
+    """
+    n = len(A)
+    AM = A[:]
+    for focus_diagonal in range(n):
+        for i in range(focus_diagonal + 1, n):
+            if AM[focus_diagonal][focus_diagonal] == 0:
+                AM[focus_diagonal][focus_diagonal] == 1.0e-18
+            row_scaler = AM[i][focus_diagonal] / AM[focus_diagonal][focus_diagonal]
+            for j in range(n):
+                AM[i][j] = AM[i][j] - row_scaler * AM[focus_diagonal][j]
+    determinant = 1.0
+    for i in range(n):
+        determinant *= AM[i][i]
+    return determinant
+
+
+def prob_calc(seq_dict):
+    """
+    Calculate probability of each player.
+
+    :param seq_dict: players sequences
+    :type seq_dict: dict
+    :return: players win probabilties as a dict.
+    """
+    prob_dic = {}
+    C = []
+    p_seq = lambda seq: 1 / 2 ** len(seq)
+    names = list(seq_dict.keys())
+    for i in range(len(names)):
+        Ai = seq_dict[str(names[i])]
+        C_row = []
+        for j in range(len(names)):
+            Aj = seq_dict[str(names[j])]
+            Wij = 0
+            for k in range(1, min(len(Ai), len(Aj)) + 1):
+                if Ai[:k] == Aj[len(Aj) - k:]:
+                    Wij += p_seq(Ai[k:])
+            C_row.append(Wij)
+        C.append(C_row)
+    C = np.array(C)
+    detCj_dic = {}
+    for j in range(len(names)):
+        column_j = C[:, j]
+        C[:, j] = 1
+        detCj_dic[names[j]] = det(C)
+        C[:, j] = column_j
+    sum_det = sum(detCj_dic.values())
+    for name in names:
+        prob_dic[name] = detCj_dic[name] / sum_det
+    return prob_dic
+
+
+def print_prob(prob_dic):
+    """
+    Print win probabilities of players.
+
+    :param prob_dic: win probability dictionary
+    :type prob_dic: dict
+    :return: None
+    """
+    sorted_probs = sorted(
+        prob_dic.items(), key=lambda x: (
+            x[1], x[0]), reverse=True)
+    name_max_length = max(map(len, prob_dic.keys()))
+    print("Wining Probability : ")
+    for item in sorted_probs:
+        prob = item[1]
+        name = item[0]
+        space_name = (name_max_length - len(name) + 5) * " "
+        print(name + space_name + "{:0.5f}%".format(prob * 100))
+    if len(set(prob_dic.values())) > 1:
+        print("Winner Should be : {}".format(sorted_probs[0][0]))
+    else:
```

Comment:
This else is not necessary in `prob_calc` function (remove `Tie` condition)

Reply:
I think it may be useful when two player have similar chance to win like "HTH" and "THT".
In this case probability should predict tie for the game.

Reply:
You are right about **two players**.
I'm not sure this condition is possible or not, but consider it :
Player 1 : 34 %
Player 2 : 34 %
Player 3 : 32 %

?!

Reply:
I think it can be called tie in a particular way.

Reply:
Oh.
Got your point now.